Minimize blocking in Filesystem.getCachedDiskUsage

cyberkitsune opened the pull request:
This resolves pterodactyl/panel#2287

When the disk size cache expires BUT another process is already refreshing it, just return the stale size instead of locking.

This prevents several other things (namely console updates) from freezing if the disk size process is taking ages.

I added an argument to skip the cache for some processes that we're ok with always blocking (like server startup)

cyberkitsune committed:
Return last cached disk size if we're currently sizing a disk

cyberkitsune committed:
Sometimes we never want to use a stale / zero value for disk usage

cyberkitsune committed:
missed a spot

cyberkitsune commented:
Don't merge quite yet, covering some additional bases. Feedback welcome still however

cyberkitsune committed:
Fix this still blocking sometimes

cyberkitsune committed:
remove redundant newline

cyberkitsune commented:
This is behaving correctly for me now.

matthewpi commented on server/filesystem.go:
```diff
@@ -43,8 +43,9 @@ func IsPathResolutionError(err error) bool {
-	lastLookupTime time.Time
-	diskUsage      int64
+	lastLookupTime   time.Time
+	lookupInProgress bool `default:"false"`
```
There is no need for the `default:"false"`, it doesn't do anything.  Also a `bool`'s default value is `false`.

matthewpi commented on server/filesystem.go:
```diff
@@ -43,8 +43,9 @@ func IsPathResolutionError(err error) bool {
 type Filesystem struct {
 	mu sync.RWMutex
```
This mutex doesn't seem to be used anywhere else, it could probably be switched to a `sync.Mutex`.

matthewpi commented on server/filesystem.go:
```diff
@@ -238,20 +239,35 @@ func (fs *Filesystem) HasSpaceAvailable() bool {
+
+	// Expire the cache after 2.5 minutes, and used the last cached value if we're currently sizing.
+	if !avoidCache && (fs.lastLookupTime.After(time.Now().Add(time.Second*-150)) || fs.lookupInProgress) {
+		return fs.diskUsage, nil
```
Use `atomic.LoadInt64` when working with `diskUsage` as it is set atomically.

matthewpi replied:
Also reading `fs.lookupInProgress` could cause a race condition as it is set with a lock but this function doesn't use one, it would probably be best to make it an atomic.

matthewpi commented on server/filesystem.go:
```diff
@@ -238,20 +239,35 @@ func (fs *Filesystem) HasSpaceAvailable() bool {
+	// otherwise trigger the update in the background and return stale value
+	go fs.updateCachedDiskUsage()
+
+	return fs.diskUsage, nil
```
Use `atomic.LoadInt64` when working with `diskUsage` as it is set atomically.

DaneEveritt commented on server/filesystem.go:
```diff
@@ -238,20 +239,35 @@ func (fs *Filesystem) HasSpaceAvailable() bool {
+func (fs *Filesystem) getCachedDiskUsage(avoidCache bool) (int64, error) {
+
+	// Expire the cache after 2.5 minutes, and used the last cached value if we're currently sizing.
+	if !avoidCache && (fs.lastLookupTime.After(time.Now().Add(time.Second*-150)) || fs.lookupInProgress) {
```
The `avoidCache` check here is unnecessary and actually causes more problems because it will fall through to the `updateCachedDiskUsage()` call even if the value in the cache is still valid, right? This should also ideally explain why this logic even exists, since the lookup time is clear but the avoid cache and lookup in progress checks aren't as intuitive.

DaneEveritt commented on server/filesystem.go:
```diff
@@ -238,20 +239,35 @@ func (fs *Filesystem) HasSpaceAvailable() bool {
 // excessive IO usage. We will only walk the filesystem and determine the size of the directory if there
 // is no longer a cached value.
-func (fs *Filesystem) getCachedDiskUsage() (int64, error) {
+func (fs *Filesystem) getCachedDiskUsage(avoidCache bool) (int64, error) {
```
I don't think `avoidCache` is a good variable name to describe what it is actually doing here. Perhaps `blockIfStale` or something? Ideally I'd prefer if it didn't even have a parameter being passed through, it feels like something we _should_ be able to check for on our own.

There is always the fun `(...block bool)` trick where we can default to `false` unless someone explicitly calls this with `true`. I'm just worried it is not going to be intuitive to people _how_ to call this function or when they should or should not be passing `true` or `false` if that makes sense.

DaneEveritt replied:
Thinking about this more, I'd like if it blocked by default, and only was a non-blocking call in specific cases. Most all of the code was written around that assumption anyways, so making it blocking unless specifically called as a non-blocking operation would make the most sense.

cyberkitsune replied:
I renamed it to "nonBlocking" and described briefly why it exists in the function comments, would that work? :)

DaneEveritt commented on server/filesystem.go:
```diff
@@ -265,6 +281,8 @@ func (fs *Filesystem) getCachedDiskUsage() (int64, error) {
 	fs.lastLookupTime = time.Now()
 	atomic.StoreInt64(&fs.diskUsage, size)
 
+	fs.lookupInProgress = false
```
For safety, this should probably be a deferred call up where it is set so if this function ever changes or manages to exit early down the road we don't forget to set this back to false.

DaneEveritt commented on server/loader.go:
```diff
@@ -128,7 +128,7 @@ func FromConfiguration(data *api.ServerConfigurationResponse) (*Server, error) {
 	// If the server's data directory exists, force disk usage calculation.
 	if _, err := os.Stat(s.Filesystem.Path()); err == nil {
-		go s.Filesystem.HasSpaceAvailable()
+		go s.Filesystem.HasSpaceAvailable(true)
```
This doesn't need to be `true`, it is "forcing" calculation simply so there is a value returned, it doesn't need to be the most accurate value. In this case, we can pass false and not even put this in a go-routine.

cyberkitsune committed:
this is false by default anyway

cyberkitsune committed:
this can just be a mutex

cyberkitsune committed:
make lookupInProgress atomic

there are no atomic booleans, so we make this an int32 instead. (This saves memory over an int64)

cyberkitsune committed:
load fs.diskUsage atomically

cyberkitsune committed:
Move defer to where it makes sense

cyberkitsune committed:
cleaned up getCachedDiskUsage and changed the logic to be block by default

cyberkitsune committed:
remove unused local

cyberkitsune committed:
Change avoidCache -> nonBlocking

cyberkitsune committed:
this doesn't need to be a goroutine anymore

as now this call is non-blocking :)

cyberkitsune committed:
Merge branch 'develop' into no-disk-block

DaneEveritt committed:
Merge branch 'develop' into pr/53

DaneEveritt committed:
Update sftp.go
